connectors/google/google-gcs/src/main/java/io/camunda/connector/google/gcs/model/core/ObjectStorageExecutor.java

```diff
@@ -0,0 +1,125 @@
+import io.camunda.connector.google.gcs.model.request.ObjectStorageRequest;
+import io.camunda.connector.google.gcs.model.request.UploadObject;
+import io.camunda.connector.google.gcs.model.response.DownloadResponse;
+import io.camunda.connector.google.gcs.model.response.DownloadResponse.DocumentContent;
+import io.camunda.connector.google.gcs.model.response.UploadResponse;
+import io.camunda.document.Document;
+import io.camunda.document.store.DocumentCreationRequest;
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.util.Optional;
+import java.util.function.Function;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class ObjectStorageExecutor {
+
+  private static final Logger log = LoggerFactory.getLogger(ObjectStorageExecutor.class);
+
+  private final Function<DocumentCreationRequest, Document> createDocument;
+  private final Authentication authentication;
+
+  public ObjectStorageExecutor(
+      Authentication authentication, Function<DocumentCreationRequest, Document> createDocument) {
+    this.authentication = authentication;
+    this.createDocument = createDocument;
+  }
+
+  public static ObjectStorageExecutor create(
+      ObjectStorageRequest objectStorageRequest,
+      Function<DocumentCreationRequest, Document> createDocument) {
+    return new ObjectStorageExecutor(objectStorageRequest.getAuthentication(), createDocument);
+  }
+
+  public Object execute(ObjectStorageOperation objectStorageOperation) {
+    return switch (objectStorageOperation) {
+      case DownloadObject downloadObject -> download(downloadObject);
+      case UploadObject uploadObject -> upload(uploadObject);
+    };
+  }
+
+  private UploadResponse upload(UploadObject uploadObject) {
+    String fileName =
+        uploadObject.fileName() != null && !uploadObject.fileName().isEmpty()
+            ? uploadObject.fileName()
+            : uploadObject.document().metadata().getFileName();
+    BlobId blobId = BlobId.of(uploadObject.bucket(), fileName);
+    BlobInfo blobInfo = BlobInfo.newBuilder(blobId).build();
+
+    try {
+      ServiceAccountCredentials sac =
+          ServiceAccountCredentials.fromStream(
+              new ByteArrayInputStream(
+                  authentication.getJsonKey().getBytes(StandardCharsets.UTF_8)));
+      Storage storage =
+          StorageOptions.newBuilder()
+              .setProjectId(uploadObject.project())
+              .setCredentials(sac)
+              .build()
+              .getService();
+
+      Blob blob = storage.createFrom(blobInfo, uploadObject.document().asInputStream());
+      return new UploadResponse(blob.getBucket(), blob.getName());
+    } catch (IOException ioException) {
+      log.error(
+          "GCS: Failed to upload file {} to bucket {}: {}",
+          fileName,
+          uploadObject.bucket(),
+          ioException.getMessage());
```

Comment:
nit: consider including the actual exception as the last param to make sure stack trace is also logged